### ISSUE TITLE
SQL Server Connection Hangs when Timeout is Set on Connection String

### ISSUE THREAD

**Issue by thiagobucca** [mono/mono#13065]
Hello, i'm experiencing a weird issue using latest version of mono with Visual Studio 8 (Preview) on macos (Latest)

I have a simple application that fetch some data from Sql Server Azure DB using SqlConnection from System.Data.SqlClient and whenever my connection string has the parameter "Connection Timeout=0" the connection just hangs and gets stuck forever. If i remove this flag from my connection string, the connection passes instantly.

I tried the same scenario using Windows 10 and latest Visual Studio and the issue is not present.

Anyone else having this problem?

## Steps to Reproduce

1. Make a simple .NET SQL SERVER Connection using SqlConnection, with a connection string that uses the parameter "Connection Timeout=0"

Connection string example:

Server=tcp:myinstance.net,1433;Initial Catalog=mydb;Persist Security Info=False;User ID=myuser;Password=mypassword;MultipleActiveResultSets=False;Encrypt=True;TrustServerCertificate=False;**Connection Timeout=0**

### Current Behavior

Visual Studio Hangs with connection forever.

### Expected Behavior

The connection works without any issues.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

5.18.0.248

**Comment by MaximLipnin** [comment]
Did you try to set non-zero connection timeout?

Btw [MS doc](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlconnection.connectiontimeout?view=netframework-4.7.2#System_Data_SqlClient_SqlConnection_ConnectionTimeout) says: `A value of 0 indicates no limit, and should be avoided in a ConnectionString because an attempt to connect waits indefinitely.`

I had no luck reproducing it using 5.18 and latest master with some remote instance of MS SQL Server with tcp/ip configured. I'll try to set up some Azure SQL Database instance and check one more time.

**Comment by MaximLipnin** [comment]
I was able to reproduce this issue with Azure SQL DB instance. BTW Azure SQL DB connection string has connection timeout = `30` by default.

**Comment by thiagobucca** [comment]
@MaximLipnin Yes, the issue only occurred to my on Azure SQL DB instances.

**Comment by MaximLipnin** [comment]
@thiagobucca if it's possible please use non-zero connection timeout to mitigate the issue.

**Comment by thiagobucca** [comment]
That's exactly my problem, i work in an enviroment that i can't touch the connection strings, i know i can just override this later, but then i have to undo this every time and this can cause issues if i forget. I'll wait for the fix, using a VM on Windows until that happens.

**Comment by MaximLipnin** [comment]
Checked this issue on latest mono (6.1). It can occur from time to time 
I use this snippet:
```
using System;
using System.Data.SqlClient;

namespace MyTest
{
    public class Repro
    {
        public static void Main()
        {
            OpenSqlConnection();
        }

        private static void OpenSqlConnection()
        {
            string connectionString = GetConnectionString();
            using (SqlConnection connection = new SqlConnection())
            {
                connection.ConnectionString = connectionString;
                connection.Open();
                Console.WriteLine("State: {0}", connection.State);
                Console.WriteLine("ConnectionString: {0}",  connection.ConnectionString);
            }
        }

        static private string GetConnectionString()
        {
            return "<some connection string stuff>;Connection Timeout=0;";
        }
    }
}

```

After some debugging I've got this error:

```
{System.Data.SqlClient.SqlException (0x80131904): Snix_Connect (provider: SNI_PN7, error: 35 - SNI_ERROR_35)

Snix_Connect (provider: SNI_PN7, error: 35 - SNI_ERROR_35) ---> System.ObjectDisposedException: Cannot access a disposed object.

Object name: 'System.Net.Sockets.Socket'.

  at System.Net.Sockets.Socket.ThrowIfDisposedAndClosed () [0x00010] in /Users/admin/dev/mono/mcs/class/System/System.Net.Sockets/Socket.cs:2729

  at System.Net.Sockets.Socket.set_NoDelay (System.Boolean value) [0x00000] in /Users/admin/dev/mono/mcs/class/System/System.Net.Sockets/Socket.cs:354

  at System.Data.SqlClient.SNI.SNITCPHandle..ctor (System.String serverName, System.Int32 port, System.Int64 timerExpire, System.Object callbackObject, System.Boolean parallel) [0x00138] in /Users/admin/dev/mono/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs:159

  at System.Data.SqlClient.SqlInternalConnection.OnError (System.Data.SqlClient.SqlException exception, System.Boolean breakConnection, System.Action`1[T] wrapCloseInAction) [0x00027] in /Users/admin/dev/mono/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnection.cs:618

  at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning (System.Data.SqlClient.TdsParserStateObject stateObj, System.Boolean callerHasConnectionLock, System.Boolean asyncClose) [0x00152] in /Users/admin/dev/mono/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs:1132

  at System.Data.SqlClient.TdsParser.Connect (System.Data.SqlClient.ServerInfo serverInfo, System.Data.SqlClient.SqlInternalConnectionTds connHandler, System.Boolean ignoreSniOpenTimeout, System.Int64 timerExpire, System.Boolean encrypt, System.Boolean trustServerCert, System.Boolean integratedSecurity, System.Boolean withFailover) [0x000ea] in /Users/admin/dev/mono/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs:334

  at System.Data.SqlClient.SqlInternalConnectionTds.AttemptOneLogin (System.Data.SqlClient.ServerInfo serverInfo, System.String newPassword, System.Security.SecureString newSecurePassword, System.Boolean ignoreSniOpenTimeout, System.Data.ProviderBase.TimeoutTimer timeout, System.Boolean withFailover) [0x00018] in /Users/admin/dev/mono/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnectionTds.cs:1647

  at System.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover (System.Data.SqlClient.ServerInfo serverInfo, System.String newPassword, System.Security.SecureString newSecurePassword, System.Boolean redirectedUserInstance, System.Data.SqlClient.SqlConnectionString connectionOptions, System.Data.SqlClient.SqlCredential credential, System.Data.ProviderBase.TimeoutTimer timeout) [0x000af] in /Users/admin/dev/mono/external/corefx/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnectionTds.cs:1336
```
Possibly it leads to some infinite loop of connection tries in case of zero-timeout.

**Comment by thiagobucca** [comment]
Don't know if it helps somehow, but sometimes, if you put a break point in connection.Open() method and press F11 to enter inside statement, it releases from infinite loop, that what i was doing before starting this thread.

**Comment by marek-safar** [comment]
@MaximLipnin could you please test this sample with .NET 2.2 if it works reliably there

**Comment by marek-safar** [comment]
@afsanehr, @tarikulsabbir, @Gary-Zh , @david-engel does this problem sound familiar?

**Comment by AfsanehR** [comment]
@marek-safar We haven't heard about this issue before. If it can be confirmed that the issue happens on a .NetCore application that uses SqlClient only, please open an issue on corefx repository and we will investigate.

**Comment by MaximLipnin** [comment]
@marek-safar There is no repro on macOS with .Net Core 2.2. I'll try to apply https://github.com/dotnet/corefx/pull/33024

**Comment by MaximLipnin** [comment]
@baulig Could you help here? Probably you have already faced similar issues...

**Comment by steveisok** [comment]
@MaximLipnin It's my understanding that the issue you linked will fix this issue.  Please confirm.  Since it's out of our camp, I'm going to remove it from the 2019-02 milestone.

**Comment by MaximLipnin** [comment]
@steveisok The fix from corefx issue 58 worked for me at that moment.

**Comment by akoeplinger** [comment]
The fix landed in mono master and 2019-06 (Mono 6.4.x) branches.
